registry/storage/layerwriter.go

```diff
@@ -83,37 +86,213 @@ func (lw *layerWriter) Cancel() error {
 	return nil
 }
 
-// validateLayer checks the layer data against the digest, returning an error
-// if it does not match. The canonical digest is returned.
-func (lw *layerWriter) validateLayer(dgst digest.Digest) (digest.Digest, error) {
-	digestVerifier, err := digest.NewDigestVerifier(dgst)
+func (lw *layerWriter) Write(p []byte) (int, error) {
+	return io.MultiWriter(lw.resumableDigester, &lw.bufferedFileWriter).Write(p)
+}
+
+func (lw *layerWriter) ReadFrom(r io.Reader) (n int64, err error) {
+	return lw.bufferedFileWriter.ReadFrom(io.TeeReader(r, lw.resumableDigester))
+}
+
+func (lw *layerWriter) Seek(offset int64, whence int) (int64, error) {
+	newOffset, err := lw.bufferedFileWriter.Seek(offset, whence)
 	if err != nil {
-		return "", err
+		return newOffset, err
+	}
+
+	if newOffset != int64(lw.resumableDigester.Len()) {
+		if err := lw.resumeHashAt(newOffset); err != nil {
```

Comment:
It might be better if this was lazy. Not sure if this is an easy change to make but just have seek invalidate the resumable hash state, then only repopulate it if another write is issued.